Add progressive mode

ssbarnea opened the pull request:
Adds a new option --progressive that ease adoption of the linter as it will only report failure if it detects an increase number of violations since previous commit.

An example of the output when run in progressive mode can be seen below:
```
WARNING  Total violation(s) reduced from 165 to 0 since previous commit. Will mark result as success
WARNING  Marked 99 previously known violation(s) as ignored due to progressive mode.
WARNING  Listing 99 violation(s) marked as ignored, likely already known
```

MarkusTeufelberger commented on lib/ansiblelint/__main__.py:
```diff
@@ -151,23 +154,28 @@ def main() -> int:
+    mark_as_success = False
+    if matches and options.progressive:
+        _logger.info(
+            "Matches found, running again on previos revision in order to detect regressions")
```
```suggestion
            "Matches found, running again on previous revision in order to detect regressions")
```

MarkusTeufelberger commented on lib/ansiblelint/__main__.py:
```diff
@@ -151,23 +154,28 @@ def main() -> int:
+            # remove old matches from current list
+            if len(old_matches) > len(matches):
+                _logger.warning(
+                    "Total violation(s) reducted from %s to %s since previous "
```
```suggestion
                    "Total violation(s) reduced from %s to %s since previous "
```

MarkusTeufelberger commented on lib/ansiblelint/__main__.py:
```diff
@@ -177,12 +185,46 @@ def main() -> int:
+
+@contextmanager
+def _previous_revision():
+    """Create or updates a temporary workdir containing the previous revious."""
```
```suggestion
    """Create or updatesa temporary workdir containing the previous revision."""
```

MarkusTeufelberger commented:
If I have 5 errors of type A, fix them in one commit but create 4 new errors of type B in that fix, will this fail (since I introduced 4 new errors) or pass (since only 4 errors are now in the repository, even though they are new)?

ssbarnea commented:
> If I have 5 errors of type A, fix them in one commit but create 4 new errors of type B in that fix, will this fail (since I introduced 4 new errors) or pass (since only 4 errors are now in the repository, even though they are new)?

I rewrote a lot of the implementation and now it will always fail if newer violations are introduced. Take a look and more important, test locally.

geerlingguy commented:
This is an interesting feature, for sure—though the first run of Ansible Lint on an existing repo would fail, correct? Or I guess if it runs against a commit that had no new violations compared to the previous commit, then it wouldn't fail.

It would be useful for those who are adopting Ansible Lint on a repo that has a lot of history and needs a lot of cleanup.

albinvass commented on lib/ansiblelint/cli.py:
```diff
@@ -111,6 +111,12 @@ def get_cli_parser() -> argparse.ArgumentParser:
+                        action='store_true',
+                        help="Return success if it detects a reduction in number"
+                        " of violations compared with previous git commit. This "
+                        "feature works only on git repository clones.")
```
> This feature only works in git repositories
